### PR TITLE
Add ctkVTKDiscretizableColorTransferWidget

### PR DESCRIPTION
**Pull request by LucasGandel**
Add a new widget to preview and edit a color transfer function given as
a vtkScalarsToColors.
The widget owns an instance of vtkScalarsToColorsContextItem that can be
added to a vtkContextView. vtkScalarsToColorsContextItem holds three
different charts to represent the information associated to the color transfer
function :
  - vtkDiscretizableColorTransferFunctionChart: Uses
  vtkControlPointsItem to edit the color transfer function

  - vtkScalarsToColorsPreviewChart: Preview the current color transfer
  function.

  -vtkScalarsToColorsHistogramChart: Display the histogram of an input
  image while applying the color transfer function to the histogram.

The widget also owns an instance of vtkScalarsToColorsComboBox to access
different presets.

image

### CODE REVIEW

**Comment by LucasGandel** [comment]
@finetjul

File: Libs/Visualization/VTK/Core/CMakeLists.txt [hunk 1]
```diff
@@ -18,7 +18,6 @@ endif()
 
 set(KIT_export_directive "CTK_VISUALIZATION_VTK_CORE_EXPORT")
 
-
```
Comment (finetjul) [inline Libs/Visualization/VTK/Core/CMakeLists.txt]: to remove

File: Libs/Visualization/VTK/Core/CMakeLists.txt [hunk 2]
```diff
@@ -72,10 +73,14 @@ if(${VTK_VERSION_MAJOR} GREATER 5)
     vtkInteractionStyle
     vtkRenderingAnnotation
     vtkRenderingCore
+    vtkViewsContext2D
```
Comment (finetjul) [inline Libs/Visualization/VTK/Core/CMakeLists.txt]: add only if ${CTK_USE_CHARTS} is true ?

File: Libs/Visualization/VTK/Core/CMakeLists.txt [hunk 3]
```diff
@@ -72,10 +73,14 @@ if(${VTK_VERSION_MAJOR} GREATER 5)
...
     )
   if(TARGET vtkRendering${VTK_RENDERING_BACKEND})
     list(APPEND VTK_LIBRARIES vtkRendering${VTK_RENDERING_BACKEND})
   endif()
+  if(TARGET vtkRenderingContext${VTK_RENDERING_BACKEND})
+    list(APPEND VTK_LIBRARIES vtkRenderingContext${VTK_RENDERING_BACKEND})
```
Comment (finetjul) [inline Libs/Visualization/VTK/Core/CMakeLists.txt]: add only if ${CTK_USE_CHARTS} is true ?

File: Libs/Visualization/VTK/Core/ctkVTKScalarsToColorsUtils.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+#include "ctkVisualizationVTKCoreExport.h"
+
+// VTK includes
+class vtkDiscretizableColorTransferFunction;
+class vtkScalarsToColors;
+
+namespace ctk
+{
+
+  /// \ingroup Visualization_VTK_Core
```
Comment (finetjul) [inline Libs/Visualization/VTK/Core/ctkVTKScalarsToColorsUtils.h]: bad indent

File: Libs/Visualization/VTK/Core/ctkVTKScalarsToColorsUtils.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+/// Remap color transfer function into the specified range
+void CTK_VISUALIZATION_VTK_CORE_EXPORT remapColorScale(
+  vtkDiscretizableColorTransferFunction* colorTransferFunction,
+  double minRescale, double maxRescale);
+
+void CTK_VISUALIZATION_VTK_CORE_EXPORT remapColorScale(
```
Comment (finetjul) [inline Libs/Visualization/VTK/Core/ctkVTKScalarsToColorsUtils.h]: add comments

File: Libs/Visualization/VTK/Core/vtkDiscretizableColorTransferChart.cpp [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+  limitations under the License.
+
+=========================================================================*/
+#include "vtkDiscretizableColorTransferChart.h"
+
+#include "ctkVTKScalarsToColorsUtils.h"
+#include "vtkDiscretizableColorTransferControlPointsItem.h"
+#include "vtkHistogramMarker.h"
+
+#include <vtkAxis.h>
```
Comment (finetjul) [inline Libs/Visualization/VTK/Core/vtkDiscretizableColorTransferChart.cpp]: not needed

File: Libs/Visualization/VTK/Core/vtkHistogramMarker.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+=========================================================================*/
+
+#include <vtkHistogramMarker.h>
+
+#include <vtkTable.h>
+#include <vtkFloatArray.h>
```
Comment (finetjul) [inline Libs/Visualization/VTK/Core/vtkHistogramMarker.cpp]: please sort alphabetically

File: Libs/Visualization/VTK/Core/vtkHistogramMarker.cpp [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+=========================================================================*/
+
+#include <vtkHistogramMarker.h>
```
Comment (finetjul) [inline Libs/Visualization/VTK/Core/vtkHistogramMarker.cpp]: #include "vtkHistogramMarker.h"

File: Libs/Visualization/VTK/Core/vtkHistogramMarker.h [hunk 9]
```diff
@@ -18,22 +18,30 @@
...
-}
+  static vtkHistogramMarker* New();
 
-#endif
+  vtkHistogramMarker();
+  virtual ~vtkHistogramMarker();
+
+  void SetPosition(double pos);
+
+  vtkSmartPointer<vtkTable> getTable();
```
Comment (finetjul) [inline Libs/Visualization/VTK/Core/vtkHistogramMarker.h]: GetTable

File: Libs/Visualization/VTK/Core/vtkHistogramMarker.h [hunk 10]
```diff
@@ -18,22 +18,30 @@
...
+
+private:
+  vtkSmartPointer<vtkTable> m_table;
+
+};
+
+#endif /* __vtkHistogramMarker_h */
```
Comment (finetjul) [inline Libs/Visualization/VTK/Core/vtkHistogramMarker.h]: Move this class in vtkDiscretizableColorTransferControlPointsItem.cxx

**Comment by jcfr** [comment]
Thanks @LucasGandel and @finetjul  😄 

Ps: When you have a chance, could you add a screenshot on the PR. And then in the image gallery when this is integrated. Merci
See http://www.commontk.org/index.php/Documentation/ImageGallery

File: Libs/Visualization/VTK/Core/vtkDiscretizableColorTransferChart.cpp (hunk 6 above)
Comment (LucasGandel) [reply Libs/Visualization/VTK/Core/vtkDiscretizableColorTransferChart.cpp]: needed for this->GetAxis()

File: Libs/Visualization/VTK/Core/vtkHistogramMarker.h (hunk 10 above)
Comment (LucasGandel) [reply Libs/Visualization/VTK/Core/vtkHistogramMarker.h]: vtkHistogramMarker needed by vtkDiscretizableColorTransferChart. It should be defined in vtkDiscretizableColorTransferChart.cpp or stay in its own file.

**Comment by LucasGandel** [comment]
@finetjul Could you please review?

**Comment by pieper** [comment]
There were some build errors in Slicer from this:
```
/Users/pieper/slicer4/latest/Slicer/Libs/MRML/Widgets/qMRMLSceneColorTableModel.cxx:67:42: error: no member named 'scalarsToColorsImage' in namespace 'ctk'
  this->Pixmap = QPixmap::fromImage(ctk::scalarsToColorsImage( scalarsToColors, this->Pixmap.size() ));
                                    ~~~~~^
1 error generated.
```
They were resolved with these fixes:

https://github.com/Slicer/Slicer/commit/edbd121747178898b40fc4c71523c36d22b526e0

**Comment by jcfr** [comment]
Thanks @pieper

**Pull request merged**